Extension types page (#5508)
dart-lang/site-www: src/language/extension-types.md

  bool isValid() => !value.isNegative;
}
```

Members of the representation type are not available to the extension type
by default. If you want a member from the representation type to be available
to the extension type, you must write a declaration for it
in the extension type definition, like the `operator +` in `NumberE`.
You also can define new members unrelated to the representation type,
like the `i` getter and `isValid` method.

Extension types can't declare instance fields (unless [`external`][]),
or [abstract members][].

### Implements

You can optionally use the `implements` clause to introduce
a subtype relationship on an extension type, making *all* members
of the representation type available to the extension type:

```dart
extension type NumberI(int i) 
  implements int{
  // ...
}
```

An extension type can only implement its representation type
or a supertype of its representation type.

This is technically NOT an inheritance relationship
like the one `implements` introduces for classes,
but rather an [applicability][] relationship like that between an
extension method and its `on` type.
Members that apply to the supertype are applicable to the subtype,

> unless the supertype has a more specific definition for the member.

Author: @eernstg I'm not sure if this statement is correct (or worth mentioning?)

---
title: Extension types
description: Learn how to write a static-only interface for an existing type.
prevpage:
  url: /language/extension-methods
  title: Extension methods
nextpage:
  url: /language/callable-objects
  title: Callable objects
---

> An extension type is a compile time abstraction that "wraps"

Author: > Of course, please just ignore this comment if this is a discussion that you've had already a long time ago

Not at all! I wasn't sure while writing the whole page (and it comes up a lot). I probably jumped between all variations so now I can go back and fix that in an informed way. Not a pet peeve in my opinion, very useful to know :)

an existing type with a different, static-only interface.
They allow you to enforce discipline on the set of operations available
to objects of the underlying type, called the *representation type*.

> You can choose to use some members, omit others,

Author: Changed to:

> When defining the interface of an extension type,
you can choose to reuse some members of the representation type,
omit others, replace others, and add new functionality.

replace others, and add new functionality, 

> all while getting compile time errors before runtime errors.

Author: Yes that was the intention, but now I think the emphasis on "compile-time" and "static-only" in the first sentence of the same paragraph make this concept obvious enough already, so I'm removing this clause

}
```

{{site.alert.note}}
Extension types serve the same purpose as **wrapper classes**,
but don't require the creation of expensive runtime objects.
Because extension types are static-only and compiled away at runtime,
they are essentially zero cost.

[**Extension methods**]() are a static abstraction similar to extension types.
However, an extension method adds functionality to *all* instances
of its underlying type. Extension types are the opposite; 
an extension type's interface only applies to objects
declared with the extension type,
and is distinct from the interface of its underlying type by default. 
Read the [Transparency](#transparency) section for more details.
{{site.alert.end}}

## Syntax

### Declaration

Define a new extension type with the `extension type` declaration and a name,
followed by the *representation type declaration* in parenthesis:

```dart
extension type E(int i) {
  // Define set of operations
}
```

The representation type declaration specifies that 
the underlying type of extension type `E` is `int`,
and the reference to the *representation object* is named `i`.

The representation object gives access to the underlying type,

> the same way declaring a final instance variable of the underlying type

Author: I changed it to

> The representation object gives the extension type access to an object
typed as the representation type, the same way declaring the type of
a final instance variable as the underlying type would in a wrapper class.

But I'm starting to think removing the second clause might be the best solution because it's a confusing sentence no matter what. I initially included it to clarify the purpose for people who are familiar with wrapper classes, but maybe that's not valuable enough to warrant any possible confusion>

would in a wrapper class.
It's in scope in the extension type body, and
you can access it using its name as a getter:

> - Within the extension type body: `this.i`.

Author: > In fact, I think our lints will likely suggest removing the this.

Just checked, that is true. Thanks

> - Within the extension type body: `this.i`.

Author: @lrhn would it make sense to say:

> - Within the extension type body: `i` or `this.i` in a constructor.

- Outside with a property extraction: `e.i` (where `e` is the object of an extension type). 

Extension types declarations can also include [type parameters](generics)
just like classes:

```dart
extension type E<T>(int i) {
  // ...
}
```

### Constructors

You can optionally declare constructors in an extension type's body.

> A constructor must be named, because the the representation declaration

Reviewer: ```suggestion
A constructor must be named, because the representation declaration
```

> A constructor must be named, because the the representation declaration

Author: Is it something to do with the `Name._` syntax that allows you to name the primary constructor, or is that just an arbitrary detail you and @lrhn happened to include in your named primary constructor examples? (I don't know what the `._` does there)

> A constructor must be named, because the the representation declaration

Reviewer 2: Every constructor has a name. In a class/enum/extension-type/... declaration named `D`, a constructor can have the name `D`. Other constructors must then have a different name (in order to avoid name clashes), and we can use anything of the form `'D' '.' <identifier>` to do that, for example, `D.myName`.

With that in mind, some people like to say that the constructor named `D` is "unnamed", "nameless", "anonymous", or something like that, and they say that constructors with a name of the form `D.myName` are "named".

I prefer to just talk about the constructor and its name, never referring to any of them being "named" or "unnamed".

So the primary constructor has the name `E._` when it's declared as in `extension type E._(...) ...`, and it has the name `E` when it is declared as in `extension type E(...) ...`.

The reason why both Lasse and I chose `_` is probably that this is the shortest possible private name, and the whole point of using `extension type E._(...) ...` is that we _don't_ want to provide a constructor which is unconditionally creating an expression of the extension type based on an expression of the representation type.

In other words, we don't want to leave the door completely open for an introduction of the extension type, we want to have more control. For example:

```dart
int _currentIdNumber = 0;

extension type IdNumber._(int _value) {
  IdNumber(): this(_currentIdNumber++);
}
```

The intended and well-behaved approach is to call the public constructor with the name `IdNumber`, and that one does not accept any arguments because it is intended to simply use the next natural number that hasn't been used so far. It is still possible to do `42 as IdNumber`, but we should all know that this is a bad idea, and the fact that we can make the primary constructor private helps communicating this rule.

> A constructor must be named, because the the representation declaration

Author: Thank you very much for the detailed explanation!

already acts as an unnamed constructor for the extension type

> that initializes the represenation object.

Author: @lrhn Sorry, I can't tell if this example is showing something different from the basic example in @eernstg's comment on the same thing. I mean there are obvious differences like the type parameters and making it `const`, but are those details relevant to naming the representation declaration, or did you just happen to include them because they were part of the actual extension type you wrote? Do you think we should include them here?

Any additional constructors must declare the representation object's
instance variable in its initializer list:

```dart
// 'E' has a named constructor 'n',

> // as well as its built-in unnamed constructor

Author: Oh I see. This seems like something to add to the declaration syntax section, that would clarify a lot of things. I will come back to this

extension type E(int i) {
  E.n(this.i, String foo);
}

void main() {
  var named = E.n(3, "Hello");
  var unnamed = E(4);

> }

Author: Thank you very much!

```

### Members

Declare members in the body of an extension type to define its interface. 
Extension type member declarations are identical to class members.
Members can be methods, getters, setters, or operators:

```dart
extension type NumberE(int value) {
  // Operator:
  NumberE operator +(NumberE other) =>
      NumberE(value + other.value);
  // Getter:

>   NumberE get i => value;

Reviewer 2: Perhaps?:
```suggestion
  NumberE get i => this;
```

We could also return `NumberE(value)`, but `this` is already available, and it is the same object as `value`, and it has the required static type, so why not use it?

Author: Sgtm, thanks!

  // Method:
  bool isValid() => !value.isNegative;
}
```

> Members of the representation type are not available to the extension type

Author: Used the longer form, thanks

by default. If you want a member from the representation type to be available
to the extension type, you must write a declaration for it
in the extension type definition, like the `operator +` in `NumberE`.
You also can define new members unrelated to the representation type,
like the `i` getter and `isValid` method.

> Extension types can't declare instance fields (unless [`external`][]),

Author: Changed to "instance variables"!

or [abstract members][].

### Implements

You can optionally use the `implements` clause to introduce
a subtype relationship on an extension type, making *all* members
of the representation type available to the extension type:

```dart
extension type NumberI(int i) 
  implements int{
  // ...
}
```

An extension type can only implement its representation type

> or a supertype of its representation type.

Author: I'm ok with adding that

This is technically NOT an inheritance relationship
like the one `implements` introduces for classes,
but rather an [applicability][] relationship like that between an
extension method and its `on` type.
Members that apply to the supertype are applicable to the subtype,
unless the supertype has a more specific definition for the member.

#### `@redeclare`

> Replacing a method that shares a name with a member of the supertype

Author: Great catch, thanks!

> is not an override relationship for extension types, but a *redeclaration*.

Author: I like the explanation but yes probably too much to say here :/ (I still added some of it though)

Where are annotations typically defined? It would be great to have this explanation wherever that it, and point to it from here.

Use the `@redeclare` annotation to tell the compiler you are *knowingly* choosing
to use the same name as a supertype's member definition, not just accidentally
using the same name:

```dart
class C {
  void m() {}
}

extension type E(C c) implements C {
  @redeclare
  void m() {}
}
```

If you don't use the annotation, the compiler will choose the [more specific][]

> definition wherever the name is invoked, which might not be the behavior you expect.

Author: Oh, ok. I was definitely confusing the lint and the annotation. Thanks!

> definition wherever the name is invoked, which might not be the behavior you expect.

Author: I had linked to [this section](https://github.com/dart-lang/language/blob/main/accepted/2.7/static-extension-methods/feature-specification.md#specificity) in a previous commit, should I put that back?

I'm starting to think it'd be best to just point to `@redeclare` / `annotate_redeclares` documentation from the "Implements" section, and skip the dedicated `@redelcare` section altogether. The lint documentation isn't very descriptive though. Fixing that could be one option. But where are annotations documented (if at all)? I can't find anything for `@redeclare`

## Usage

To use an extension type, create an instance the same as you would with a class:

```dart
extension type NumberE(int value) {
  NumberE operator +(NumberE other) =>
      NumberE(value + other.value);

  NumberE get i => value;
  bool isValid() => !value.isNegative;
}

void testE() { 
  var num = NumberE(1);
}
```

Then you can invoke members on the object as you would with a class object.

> Any attempt to treat the extension type object as an object

Reviewer 3: "Any attempt" is a little ambitious.
I'd usually include `o as RepresentationType` in ways to treat an object as `RepresentationType, and that'll work fine.
Is this trying to say: *Can't assign extension type to representation type, or invoke representation type members on extension type, unless the extension type implements the representation type. Can't assign representation type to extension type ever (unless representation type is `dynamic`).

If that's the point, then I think it has been mostly made above already, but summarizing can be useful.

Maybe:
> Subtyping and assignability works normally for extension
> types, which are (statically) treated as completely new types,
> subtypes of `Object?` and supertypes of `Never`, 
> and otherwise only related to other types through their
> `implements` clauses.
> An extension type cannot be assigned to its representation
> type, and it doesn't expose the representation type's 
> members, not unless the extension type explicitly implements
> the representation type to make itself be a subtype of the 
> representation type (see [transparency](#transparency)). 
>
> Likewise, the representation type of an extension type
> is never a subtype of the extension type, so it cannot
> be used where the extension type is needed.

Author: > If that's the point, then I think it has been mostly made above already,

That is the point, and I think you're right. There's a lot of ideas repeated in different sections of this doc I should go through and minimize.

of the representation type, or vice versa, will result in compile time errors
(unless that behavior is explicitly defined for the extension type,

> or it is [transparent](#transparency)).

Reviewer 3: Not sure how it can be explicitly defined *other* than being "transparent", so this paragraph confuses me.

Consider (if not taking the suggestion above):
> (unless the extension type implements the representation type, then it can be used as having the representation type).

For example:

```dart
void testE() { 

>   var num = NumberE(1);

Reviewer 3: Call this `num1`. Don't want to get confused with the type `num`.

Reviewer: Was thinking similar (but `n`). `num` is used as a variable name several times in the code samples.

Author: Oof, big oversight on my end. Thanks for catching!

  int num2 = NumberE(2); // Error: Can't assign 'NumberE' to 'int'.
  

>   num.isValid(); // Ok: Extension member invocation.

Reviewer 2: ```suggestion
  num.isValid(); // Ok: Extension type member invocation.
```

I think we have to reserve the phrase 'extension member' to the ones that occur in an `extension` declaration, not an `extension type` declaration.

  num.isNegative(); // Error: 'NumberE' does not define 'int' member 'isNegative'.
  
  var sum = num + num; // Ok: 'NumberE' defines '+'.
  var diff = num - num; // Error: 'NumberE' does not define 'int' member '-'.
  var diff2 = num.value - 2; // Ok: Can access representation object with reference.
  var sum2 = num + 2; // Error: Can't assign 'int' to parameter type 'NumberE'. 
  
  List<NumberE> numbers = [
    NumberE(1), 
    num.i, // Ok: 'i' getter returns type 'NumberE'.

>     1, // Error: Can't assign 'int' element to list type 'NumberE'.

Reviewer 2: Perhaps?:
```suggestion
    1, // Error: Can't assign an 'int' to the list element type 'NumberE'.
```

>     1 as NumberE, // Ok: Dynamic type cast recognizes representation type.

Reviewer 3: Not sure what "recognizes" means.

It's just OK, because it has static type `NumberE`.
It would also be (statically) OK to write `"nope" as NumberE`. That would fail at runtime, but in an example with multiple compile-time errors, runtime is a foreign country. I wouldn't expect the program to be about runtime behavior at all.

(Maybe split the example into two: One with errors, and one with correct usages. Then the reader will expect the correct usages to be *runtime* correct too.)

Author: Ok, it sounds like the `as` cast example just doesn't fit here so I'll remove it

  ];
}
```

> #### Transparency

Author: If I'm understanding correctly (from this and your other comments on this section) would it be better to say:

"An extension type that implements its representation type, or a supertype of its representation type, can be _thought of_ as transparent, because it provides a kind of overarching access to underlying type members (at least _somewhere_ in the hierarchy / either those of the direct representation type or somewhere higher in the type hierarchy) that is otherwise _not_ the default for the intrinsic isolated nature of extension type interfaces"?

Not that that's exactly what I'll write, but is my understanding correct?

> #### Transparency

Reviewer 2: > It's also possible to have extension types that implement a _supertype_ of their representation type.

True. I did not push on emphasizing this point because I think it's a natural refinement. The crucial message to communicate is that there are two equally valid, but very different, ways to use an extension type: It can be used (1) to extend the interface of an existing type (adding extra members), and (2) to create an unrelated type with an unrelated interface.

An example of the former is `FancyString` that adds extra methods to `String`; we surely want to be able to use an expression of this type in all the ways that `String` supports. An example of the latter is `IdNumber` that provides an interface which is suitable for ID numbers, but allows us to use plain (and cheap) `int` objects for that purpose; in this case we definitely do not want to silently include the interface of `int` as part of the interface of `IdNumber`.

I think those two modes of operation are so different that they should be established _conceptually_, thus clearly communicating that each of them is OK, but each of them comes with its own conceptualization and software engineering rules.

It would be helpful to support this conceptualization by means of a well-chosen terminology, and 'transparent' seems fine to me. So `FancyString` is transparent because we "can see" the underlying type `String`, and `IdNumber` is 'not transparent' because it is an important part of being an ID number that it does not do all the `int` things.

On top of that, we can refine: It's possible for an extension type `E` to have a representation type `R1` and a clause `implements ... R2 ...` where `R1` is a proper subtype of `R2`, e.g., because `R1` is private and `R2` defines the part of the interface of `R1` that we wish to provide to clients. We can also have a mostly-transparent extension type that adds some new members and _adapts_ others by redeclaring the given name (e.g., in order to use stricter types on some parameters of a method, or different default values, etc). Another way to be mostly transparent is to implement a type which is a proper supertype of the representation type.

My take on the refinements was that there may or may not be room for that level of detail here. If we can help developers to firmly establish the notion of an extension type as transparent or not transparent then it won't be a problem to deal with everything in between.

> #### Transparency

Author: Ok, I think I figured out how to handle this (commit coming up!).

1. I'll include Lasse's points about the different potential `implements` scenarios with examples in the "Implements" section. Since that section comes first, this will hopefully set the precedent that we are *not* implying implementing the representation type or nothing at all are your only options. 

2. Then later on in the "Usage" section, I'll describe Erik's crucial use cases, transparent and non-transparent. I'll include information about the other `implements` situations in the "transparent" case section to not gatekeep the scenario too much. I'll generally try not to be too strict about "transparency" as a title, but more just "you can _think_ of these as transparent because..." (also based on Erik's comment here). 

Thanks for the attention to detail!

An extension type that [implements](#implements) its representation type
is considered *transparent* because it gives the extension type access

> to *all* members of the representation type. Otherwise,

Reviewer 2: ```suggestion
to *all* members of the representation type that aren't redeclared. Otherwise,
```

only the members declared in the extension type body are available to it
(considered a [*somewhat protected*](#type-considerations) extension type).

```dart
extension type NumberT(int value) 
  implements int {
  // Doesn't explicitly declare any members of 'int'.
  NumberT get i => NumberT(value);
}

void main () {
  var num = NumberT(1);
  

>   // All OK: Representation type interface is available to extension type:

Reviewer 2: Perhaps?:
```suggestion
  // All OK: `implements int` allows using `int` members on the extension type:
```

The point is that the decision is made based on the type in the implements clause. If we had written `implements num` then we would not be able to invoke `isEven` on an expression of type `NumberT`, even in the case where the representation type is `int`.

There is an extra constraint (that you already mention elsewhere): If we have `implements T` where `T` is a non-extension type then it must be true that the representation type `R` is a subtype of `T`.

It's probably going to be rather common that an extension type uses the same non-extension type `T` as (1) the representation type and (2) as a superinterface (as in `implements ...T...`), but there are other possibilities.

  int numT = NumberT(2);
  num.i - num; numT + num; num - 1; 2 + num;
  
  // Error: Extension type interface is not available to representation type:
  numT.i;
}
```

> Transparency is a conceptual decision for the extension type creator to consider.

Author: Haha good point!

A transparent extension type provides an *extended* interface to an existing type.

> You can invoke all the members of the representation type,

Reviewer 2: ```suggestion
You can invoke all the members of the representation type that aren't redeclared,
```

plus any auxillary members you declare. The new interface is available

> to instances of the extension type.

Reviewer 2: ```suggestion
to expressions whose static type is the extension type.
```

> to instances of the extension type.

Reviewer 2: It's actually about expressions and their static type. Perhaps this will do:
```suggestion
for expressions of the extension type.
```
(OK, github refuses to specify a `suggestion` here, but it's intended to replace line 267.)

The inverse to transparency is restricting all members of the representation,
providing a totally *different* interface to an existing type.
This means simply not including it in the extension type's implements clause,

> and not declaring any of its members in its definition.

Author: Thank you for clearing that up, I was also unsure :)

This is as close as you can get to the complete protection of a wrapper class.

## Type considerations

Extension types are a compile-time wrapping construct.

> So, at runtime, there is absolutely no trace of the extension type.

Reviewer 2: ```suggestion
So, at run time, there is absolutely no trace of the extension type.
```

Any type query or similar runtime expression works on the representation type.

This makes extension types an *unsafe* abstraction,
because you can always find out the representation type at runtime
and access the underlying object.

Dynamic type tests (`e is T`), casts (`e as T`),
and other runtime type queries (like `switch (e) ...` or `if (e case ...)`)
all evaluate to the underlying representation type. 
That's true when the static type of `e` is an extension type,
and when testing against an extension type (`case MyExtensionType(): ... `).

```dart
void main() {
  var num = NumberE(1);
  

>   if (num case int x) {  // True because runtime type is representation type.

Author: Thank you, that's much better

    print(x);
  }
  
  switch (num) {

>     case (int x): print(x);  // Matches because runtime type is representation type.

Reviewer 2: Maybe omit the parens:
```suggestion
    case int x: print(x);  // Matches because runtime type is representation type.
```

It's just going to make some readers speculate about why those parentheses are included, and whether it's even allowed. ;-)

  }

> }

Author: I added a lead-in to this example saying:
> Similarly, the static type of the matched value is that of the extension type
in this example:

But I don't think that's makes sense and I don't really understand the example. What are we showing here? Is it just another way to show that `NumberE` is `int` at run time? It doesn't seem like  "the static type of the matched value (is that `i`?) is `NumberE`" to me, but rather that the runtime type of `NumberE` is `int`
